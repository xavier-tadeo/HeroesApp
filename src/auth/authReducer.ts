import { types } from "../types/types";

// const state = {
//   name: "arlet",
//   logged: true
// }

type TAction = {
  type: string;
  payload: {
    name: string;
    email: string;
  };
};

export const authReducer = (state = {}, action: TAction) => {
  switch (action.type) {
    case types.login:
      return {
        ...action.payload,
        logged: true,
      };

    case types.logout:
      return {
        logged: false,
      };

    default:
      return state;
  }
};
