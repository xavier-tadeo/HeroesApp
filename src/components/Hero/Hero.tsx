import React from "react";

export const Hero = () => {
  return (
    <div>
      <h1>Hero</h1>
    </div>
  );
};
