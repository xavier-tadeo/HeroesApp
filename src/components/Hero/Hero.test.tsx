import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import ReactTestRenderer from "react-test-renderer";
import { AuthContext } from "../../auth/authContext";
import { Hero } from "./Hero";
const mockNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockNavigate,
}));

describe("Given Hero component", () => {
  const contextValue = {
    user: {
      name: "Xavier",
      logged: true,
    },
    dispatch: jest.fn(),
  };
  describe("When it renderer", () => {
    test("Then should show anything when don't have a hero", () => {
      const view = ReactTestRenderer.create(
        <AuthContext.Provider value={contextValue}>
          <MemoryRouter initialEntries={["/hero"]}>
            <Routes>
              <Route path="/hero" element={<Hero />} />
              <Route path="/" element={<h1>Don't have a Hero</h1>} />
            </Routes>
          </MemoryRouter>
        </AuthContext.Provider>
      );

      expect(view.toJSON()).toMatchSnapshot();
    });

    test("The should show hero whe it have a hero", () => {
      render(
        <AuthContext.Provider value={contextValue}>
          <MemoryRouter initialEntries={["/hero/marvel-spider"]}>
            <Routes>
              <Route path="/hero/:heroId" element={<Hero />} />
              <Route path="/" element={<h1>Don't have a Hero</h1>} />
            </Routes>
          </MemoryRouter>
        </AuthContext.Provider>
      );
      const view = ReactTestRenderer.create(
        <AuthContext.Provider value={contextValue}>
          <MemoryRouter initialEntries={["/hero/marvel-spider"]}>
            <Routes>
              <Route path="/hero/:heroId" element={<Hero />} />
              <Route path="/" element={<h1>Don't have a Hero</h1>} />
            </Routes>
          </MemoryRouter>
        </AuthContext.Provider>
      );

      const nameCharacter = screen.getByText("Spider Man");

      expect(view.toJSON()).toMatchSnapshot();
      expect(nameCharacter).toBeInTheDocument();
    });
  });
  describe("When user click the button", () => {
    test("Then should return to the before screen", () => {
      render(
        <AuthContext.Provider value={contextValue}>
          <MemoryRouter initialEntries={["/hero/marvel-spider"]}>
            <Routes>
              <Route path="/hero/:heroId" element={<Hero />} />
              <Route path="/" element={<h1>Don't have a Hero</h1>} />
            </Routes>
          </MemoryRouter>
        </AuthContext.Provider>
      );

      const button = screen.getByRole("button");
      fireEvent.click(button);

      expect(mockNavigate).toHaveBeenCalledWith(-1);
    });
  });
  describe("When no existing hero", () => {
    test("Then should return h1 with messege 'Don't have a hero'", () => {
      render(
        <AuthContext.Provider value={contextValue}>
          <MemoryRouter initialEntries={["/hero/marvel-spider98989898"]}>
            <Routes>
              <Route path="/hero/:heroId" element={<Hero />} />
              <Route path="/" element={<h1>Don't have a Hero</h1>} />
            </Routes>
          </MemoryRouter>
        </AuthContext.Provider>
      );

      const h1Expect = screen.getByText("Don't have a Hero");

      expect(h1Expect).toBeInTheDocument();
    });
  });
});
