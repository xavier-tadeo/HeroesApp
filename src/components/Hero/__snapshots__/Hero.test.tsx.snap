// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given Hero component When it renderer The should show hero whe it have a hero 1`] = `
<div
  className="row mt-5"
>
  <div
    className="col-4"
  >
    <img
      alt="Spider Man"
      className="img-thumbnail"
      src="/assets/marvel-spider.jpg"
    />
  </div>
  <div
    className="col-8"
  >
    <h3>
      Spider Man
    </h3>
    <ul
      className="list-group"
    >
      <li
        className="list-group-item"
      >
        <b>
          Alter ego:
        </b>
        Peter Parker
      </li>
      <li
        className="list-group-item"
      >
        <b>
          Publisher:
        </b>
        Marvel Comics
      </li>
      <li
        className="list-group-item"
      >
        <b>
          First Appearance:
        </b>
        Amazing Fantasy #15
      </li>
    </ul>
    <h5>
      Characters
    </h5>
    <p>
      Peter Parker
    </p>
    <button
      className="btn btn-outline-info"
      onClick={[Function]}
    >
      Return
    </button>
  </div>
</div>
`;

exports[`Given Hero component When it renderer Then should show anything when don't have a hero 1`] = `null`;
